src/polydatum/dal.py

```diff
@@ -1,20 +1,236 @@
+            # we cannot use attribute access on the dal directly because of how
+            # attribute access is deferred with DalMethodRequest objects.
+            if isinstance(service_or_method, dict):
+
+                # If the code being resolved has typo'd a service name, this
+                # could be returning something that is not a service.
+                service_or_method = service_or_method.get(path_segment.name)
+            else:
+                # second time through (and beyond), service or method is a real
+                # service or method, and we do not need to use a special case for
+                # finding the first service.
+                try:
+                    service_or_method = getattr(service_or_method, path_segment.name)
+                    if not service_or_method and callable(service_or_method):
+                        raise DalMethodError(request, path=tuple(location))
+
+                except KeyError:
+                    raise DalMethodError(request, path=tuple(location))
+        else:
+            raise DalMethodError(request, path=tuple(location))
+    return service_or_method
+
+
+class DalMethodResolverMiddleware(MethodMiddleware):
+    """
+    A Method Middleware that resolves deferred dal attribute access to
+    underlying services/methods that has to be called
+    """
+
+    def __call__(self, request: DalMethodRequest, handler: Callable):
+        """
+        Args:
+             request: Input from caller
+             handler: Downstream middleware or actual DAL method handler
+        """
+        request.dal_method = resolve(request)
+        return handler(request)
+
+
+def default_handle_dal_method(request: DalMethodRequest):
+    """
+    The default method middleware handler for a DalMethodRequester
+
+    Args:
+        request (DalMethodRequest): The method request context.
+
+    Returns: Mixed
+    """
+    assert request.dal_method, "DAL method not resolved"
+    return request.dal_method(*request.args, **request.kwargs)
+
+
+class InvalidMiddleware(Exception):
+    pass
 
 
 class DataAccessLayer(object):
     """
     Gives you access to a DataManager's services.
     """
 
-    def __init__(self, data_manager):
+    # default middleware classes need to be instantiated before being
+    # passed to the init function because the resulting object is called
+    # directly (__call__).
+    def __init__(
+        self,
+        data_manager,
+        middleware=None,
+        default_middleware=(DalMethodResolverMiddleware(),),
+        handler=default_handle_dal_method
+    ):
         self._services = {}
         self._data_manager = data_manager
+        self._handler = handler
+        self._reversed_middleware = []
+
+        middleware = middleware or []
+        if default_middleware:
+            middleware.extend(default_middleware)
+        for m in reversed(middleware):
+            # Be as helpful as we can to callers...
```

Comment:
Explain further.

I think you're saying "be helpful by instantiating an instance if the caller passes in a class".

Reply:
Yes that's it exactly. We got burned in tests by passing a class here because instead of "calling" the middlewares, we were instantiating them and it was annoying to lose a test cycle just to fix that. So - trying to be helpful here.